tests/functional/test_utilities_batch.py

```diff
@@ -639,3 +652,82 @@ def lambda_handler(event, context):
 
     # THEN raise BatchProcessingError
     assert "All records failed processing. " in str(e.value)
+
+
+def test_async_batch_processor_middleware_success_only(sqs_event_factory, async_record_handler):
+    # GIVEN
+    first_record = SQSRecord(sqs_event_factory("success"))
+    second_record = SQSRecord(sqs_event_factory("success"))
+    event = {"Records": [first_record.raw_event, second_record.raw_event]}
+
+    processor = AsyncBatchProcessor(event_type=EventType.SQS)
+
+    @async_batch_processor(record_handler=async_record_handler, processor=processor)
+    def lambda_handler(event, context):
+        return processor.response()
+
+    # WHEN
+    result = lambda_handler(event, {})
+
+    # THEN
+    assert result["batchItemFailures"] == []
+
+
+def test_async_batch_processor_middleware_with_failure(sqs_event_factory, async_record_handler):
+    # GIVEN
+    first_record = SQSRecord(sqs_event_factory("fail"))
+    second_record = SQSRecord(sqs_event_factory("success"))
+    third_record = SQSRecord(sqs_event_factory("fail"))
+    event = {"Records": [first_record.raw_event, second_record.raw_event, third_record.raw_event]}
+
+    processor = AsyncBatchProcessor(event_type=EventType.SQS)
+
+    @async_batch_processor(record_handler=async_record_handler, processor=processor)
+    def lambda_handler(event, context):
+        return processor.response()
+
+    # WHEN
+    result = lambda_handler(event, {})
+
+    # THEN
+    assert len(result["batchItemFailures"]) == 2
+
+
+def test_async_batch_processor_context_success_only(sqs_event_factory, async_record_handler):
+    # GIVEN
+    first_record = SQSRecord(sqs_event_factory("success"))
+    second_record = SQSRecord(sqs_event_factory("success"))
+    records = [first_record.raw_event, second_record.raw_event]
+    processor = AsyncBatchProcessor(event_type=EventType.SQS)
+
+    # WHEN
+    with processor(records, async_record_handler) as batch:
+        processed_messages = batch.async_process()
+
+    # THEN
+    assert processed_messages == [
+        ("success", first_record.body, first_record.raw_event),
+        ("success", second_record.body, second_record.raw_event),
+    ]
+
+    assert batch.response() == {"batchItemFailures": []}
+
```

Comment:
If you have bandwidth (or whoever maintainer frees up in January), let's add a test to prevent that regression we caught on custom processors?

For now, the simplest suffices so we can ensure we don't break customers who create custom processors.

Docs: https://awslabs.github.io/aws-lambda-powertools-python/2.4.0/utilities/batch/#create-your-own-partial-processor